drafts/two-way-binding-signals-and-query-params.md

```diff
@@ -0,0 +1,754 @@
+
+The waiting logic is achieved by awaiting this `Promise`:
+
+```ts
+private waitUntilRouterIsIdle(): Promise<unknown> {
+  return firstValueFrom(
+    this.router.events.pipe(
+      filter((e) => e instanceof NavigationEnd),
+      timeout(1_000)
+    )
+  );
+}
+```
+
+We simply use the `router.events` observable and filter until we find a `NavigationEnd` value. Then we just convert it to `Promise` using `firstValueFrom` so we can await it in the effect. Notice that there's also a `timeout` of 1s but we could think of a much more elaborated error handling in another iteration. 
+
+Once we have all the pieces ready, we made it! Now we have a nice utility that allow us to consume query params with signals. Not only that, those signals are writable and every time they are updated those values are propagated to Angular's router state and the browser url. 
+
+Let's now see the nicer solution of the anonymise feature implemented with these signals. And perhaps a few more interesting use cases: 
+
+```ts
+@Component({
+  selector: 'app-root',
+  standalone: true,
+  template: `
+    <h1>Signals Demo</h1>
+    <p>Full name: {{ fullName() }}</p>
+    <p>Age: {{ age() }}</p>
+
+    <hr />
+
+    <h2>Using Signals</h2>
+    <button (click)="anonymise()">I want to be anonymous 👤</button>
+    <button (click)="becomeJackSparrow()">Become Jack Sparrow 🏴‍☠️</button>
+    <button (click)="getOlder()">Get Older 👴</button>
+
+    <hr />
+
+    <h2>Using Angular Router</h2>
+    <button (click)="becomeHarryPotter()">Become Harry Potter 🪄</button>
+    <button (click)="getYounger()">Get Younger 👶</button>
+  `,
+})
+export class AppComponent {
+  private queryParamService = inject(QueryParamService);
+  private router = inject(Router);
+  private activatedRoute = inject(ActivatedRoute);
+
+  firstName = this.queryParamService.get('firstName');
+  lastName = this.queryParamService.get('lastName');
+  age = this.queryParamService.get('age');
+
+  fullName = computed(
+    () => `${this.firstName() ?? ''} ${this.lastName() ?? ''}`
+  );
+
+  // ---- using signals
+  anonymise() {
+    this.firstName.set(undefined);
+    this.lastName.set(undefined);
+  }
+
+  becomeJackSparrow() {
+    this.firstName.set('Jack');
+    this.lastName.set('Sparrow');
+  }
+
+  getOlder() {
+    this.age.set(+this.age() + 5);
+  }
+
+  // ---- using Angular router
+  becomeHarryPotter() {
+    const queryParams = { firstName: 'Harry', lastName: 'Potter' };
+    this.router.navigate([], { queryParams, queryParamsHandling: 'merge' });
+  }
+
+  getYounger() {
+    const queryParams = {
+      age: +this.activatedRoute.snapshot.queryParams['age'] - 5,
+    };
+    this.router.navigate([], { queryParams, queryParamsHandling: 'merge' });
```

Comment:
**question:** why aren't these methods using the Signal setters but still go directly through the router?